x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java

```diff
@@ -57,53 +66,96 @@ public class TransportGetCheckpointAction extends HandledTransportAction<Request
     }
 
-    protected void resolveIndicesAndGetCheckpoint(Task task, Request request, ActionListener<Response> listener, final ClusterState state) {
+    protected void resolveIndicesAndGetCheckpoint(
+        Task task,
+        Request request,
+        ActionListener<Response> listener,
+        final ClusterState clusterState
+    ) {
+        final String nodeId = clusterState.nodes().getLocalNode().getId();
+        final TaskId parentTaskId = new TaskId(nodeId, task.getId());
+
         // note: when security is turned on, the indices are already resolved
         // TODO: do a quick check and only resolve if necessary??
-        String[] concreteIndices = this.indexNameExpressionResolver.concreteIndexNames(state, request);
-
-        Map<String, Set<ShardId>> nodesAndShards = resolveIndicesToPrimaryShards(state, concreteIndices);
-
-        if (nodesAndShards.size() == 0) {
+        String[] concreteIndices = this.indexNameExpressionResolver.concreteIndexNames(clusterState, request);
+        Map<String, Set<ShardId>> nodesAndShards = resolveIndicesToPrimaryShards(clusterState, concreteIndices);
+        if (nodesAndShards.isEmpty()) {
             listener.onResponse(new Response(Collections.emptyMap()));
             return;
         }
 
-        new AsyncGetCheckpointsFromNodesAction(state, task, nodesAndShards, new OriginalIndices(request), request.getTimeout(), listener)
-            .start();
+        if (request.getQuery() == null) {  // If there is no query, then there is no point in filtering
+            getCheckpointsFromNodes(clusterState, task, nodesAndShards, new OriginalIndices(request), request.getTimeout(), listener);
+            return;
+        }
+
+        SearchShardsRequest searchShardsRequest = new SearchShardsRequest(
+            request.indices(),
+            SearchRequest.DEFAULT_INDICES_OPTIONS,
+            request.getQuery(),
+            null,
+            null,
+            false,
+            request.getCluster()
+        );
+        searchShardsRequest.setParentTask(parentTaskId);
+        ClientHelper.executeAsyncWithOrigin(
+            client,
+            ClientHelper.TRANSFORM_ORIGIN,
+            SearchShardsAction.INSTANCE,
+            searchShardsRequest,
+            ActionListener.wrap(searchShardsResponse -> {
+                Map<String, Set<ShardId>> filteredNodesAndShards = filterOutSkippedShards(nodesAndShards, searchShardsResponse);
+                getCheckpointsFromNodes(
+                    clusterState,
+                    task,
+                    filteredNodesAndShards,
+                    new OriginalIndices(request),
+                    request.getTimeout(),
+                    listener
+                );
+            }, e -> {
+                // search_shards API failed so we just log the error here and continue just like there was no query
+                logger.warn(Strings.format("search_shards API failed, request was: [%s]", searchShardsRequest), e);
```

Comment:
`searchShardsRequest.toString()` could print a _huge_ amount of data due to `note: when security is turned on, the indices are already resolved`. It could print the names of 10000+ indices.

I think the full request should be logged at the `trace` level. The `warn` level logging needs to avoid the risk of logging megabytes of data.

Reply:
Done.